File: transpile/giada/pizzapazza/math/Z4Rotation.js
/**
 * The rotation (angles parameters are computed in degrees,
 * rotations are computed in radians)
 *
 * @author gianpiero.di.blasi
 */
class Z4Rotation extends Z4AbstractFancifulValue {

   type = 0;

   startAngle = 0.0;

   delayed = false;

   rotationNext = 0.0;

  constructor(type) {
    super();
    this.type = type;
  }

  /**
   * Returns if the next rotation is computed on a fixed value
   *
   * @return true if the next rotation is computed on a fixed value, false
   * otherwise
   */
   isFixed() {
    return this.type === 0;
  }

  /**
   * Returns if next rotation is computed by cumulating previous rotation
   *
   * @return true if next rotation is computed by cumulating previous rotation,
   * false otherwise
   */
   isCumulative() {
    return this.type === 1;
  }

  /**
   * Checks if next rotation is computed relative to a path
   *
   * @return true if next rotation is computed relative to a path, false
   * otherwise
   */
   isRelativeToPath() {
    return this.type === 2;
  }

  /**
   * Returns the initial angle of rotation (in degrees)
   *
   * @return The initial angle of rotation (in degrees)
   */
   getStartAngle() {
    return this.startAngle;
  }

  /**
   * Sets the initial angle of rotation (in degrees)
   *
   * @param startAngle The initial angle of rotation (in degrees)
   * @return This Z4Rotation
   */
   setStartAngle(startAngle) {
    this.startAngle = startAngle;
    return this;
  }

  /**
   * Returns if the rotation has to be delayed (rotated by a PI angle)
   *
   * @return true if the returned rotation has to be delayed (rotated by a PI
   * angle), false otherwise
   */
   isDelayed() {
    return this.delayed;
  }

  /**
   * Sets if the rotation has to be delayed (rotated by a PI angle)
   *
   * @param delayed true if the rotation has to be delayed (rotated by a PI
   * angle), false otherwise
   * @return This Z4Rotation
   */
   setDelayed(delayed) {
    this.delayed = delayed;
    return this;
  }

  /**
   * Returns the next rotation
   *
   * @param tangentAngle The tangent angle (in radians)
   * @return The next rotation (in radians)
   */
   next(tangentAngle) {
    let angle = Z4Math.deg2rad(this.startAngle + super.next(0));
    switch(this.type) {
      case 0:
        return angle + (this.delayed ? Math.PI : 0);
      case 1:
        this.rotationNext += angle;
        return this.rotationNext + (this.delayed ? Math.PI : 0);
      case 2:
        return angle + tangentAngle + (this.delayed ? Math.PI : 0);
      default:
        return 0;
    }
  }

  /**
   * Computes the next side
   *
   * @param z4Point The current point
   * @param vector The tangent vector
   */
   nextSide(z4Point, vector) {
    switch(this.type) {
      case 0:
      case 1:
        z4Point.setSide(Z4Sign.POSITIVE);
        break;
      case 2:
        z4Point.setSide(vector ? vector.direction(z4Point.getZ4Vector()) : Z4Sign.RANDOM);
        break;
    }
  }

  /**
   * Returns a Z4Rotation with next rotation computed on a fixed value
   *
   * @return The Z4Rotation
   */
  static  fixed() {
    return new Z4Rotation(0);
  }

  /**
   * Returns a Z4Rotation with next rotation computed by cumulating previous
   * rotation
   *
   * @return The Z4Rotation
   */
  static  cumulative() {
    return new Z4Rotation(1);
  }

  /**
   * Returns a Z4Rotation with next rotation computed relative to a path
   *
   * @return The Z4Rotation
   */
  static  relativeToPath() {
    return new Z4Rotation(2);
  }
}

File: src/pizzapazza/math/Z4Rotation.java
package pizzapazza.math;

import simulation.js.$Object;

/**
 * The rotation (angles parameters have to be provided in degrees, rotations are
 * computed in radians)
 *
 * @author gianpiero.diblasi
 */
public class Z4Rotation implements Z4NextableWithParam<Double, Double> {

  private final double startAngle;
  private final Z4FancifulValue angle;
  private final Z4RotationBehavior behavior;
  private final boolean delayed;

  private double rotationNext;

  /**
   * Creates the object
   *
   * @param startAngle The initial angle of rotation (in degrees)
   * @param angle The angle (in degrees)
   * @param behavior The rotation behavior
   * @param delayed true if the returned rotation has to be delayed (rotated by
   * a PI angle), false otherwise
   */
  public Z4Rotation(double startAngle, Z4FancifulValue angle, Z4RotationBehavior behavior, boolean delayed) {
    super();

    this.startAngle = startAngle;
    this.delayed = delayed;
    this.behavior = behavior;
    this.angle = angle;
  }

  /**
   * Returns the initial angle of rotation (in degrees)
   *
   * @return The initial angle of rotation (in degrees)
   */
  public double getStartAngle() {
    return this.startAngle;
  }

  /**
   * Returns the angle (in degrees)
   *
   * @return The angle (in degrees)
   */
  public Z4FancifulValue getAngle() {
    return this.angle;
  }

  /**
   * Returns the rotation behavior
   *
   * @return The rotation behavior
   */
  public Z4RotationBehavior getRotationBehavior() {
    return this.behavior;
  }

  /**
   * Returns if the rotation has to be delayed (rotated by a PI angle)
   *
   * @return true if the returned rotation has to be delayed (rotated by a PI
   * angle), false otherwise
   */
  public boolean isDelayed() {
    return this.delayed;
  }

  @Override
  public Double next(Double tangentAngle) {
    double nextAngle = Z4Math.deg2rad(this.startAngle + this.angle.next());

    if (this.behavior == Z4RotationBehavior.FIXED) {
      return nextAngle + (this.delayed ? Math.PI : 0);
    } else if (this.behavior == Z4RotationBehavior.CUMULATIVE) {
      this.rotationNext += nextAngle;
      return this.rotationNext + (this.delayed ? Math.PI : 0);
    } else if (this.behavior == Z4RotationBehavior.RELATIVE_TO_PATH) {
      return nextAngle + tangentAngle + (this.delayed ? Math.PI : 0);
    } else {
      return 0.0;
    }
  }

  @Override
  public $Object toJSON() {
    $Object json = new $Object();
    json.$set("startAngle", this.startAngle);
    json.$set("angle", this.angle.toJSON());
    json.$set("behavior", this.behavior);
    json.$set("delayed", this.delayed);
    return json;
  }

  /**
   * Creates a Z4Rotation from a JSON object
   *
   * @param json The JSON object
   * @return the rotation
   */
  public static Z4Rotation fromJSON($Object json) {
    switch ("" + json.$get("behavior")) {
      case "FIXED":
        return new Z4Rotation(json.$get("startAngle"), Z4FancifulValue.fromJSON(json.$get("angle")), Z4RotationBehavior.FIXED, json.$get("delayed"));
      case "CUMULATIVE":
        return new Z4Rotation(json.$get("startAngle"), Z4FancifulValue.fromJSON(json.$get("angle")), Z4RotationBehavior.CUMULATIVE, json.$get("delayed"));
      case "RELATIVE_TO_PATH":
        return new Z4Rotation(json.$get("startAngle"), Z4FancifulValue.fromJSON(json.$get("angle")), Z4RotationBehavior.RELATIVE_TO_PATH, json.$get("delayed"));
      default:
        return null;
    }
  }
}
